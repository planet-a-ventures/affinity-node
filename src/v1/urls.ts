/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-a-specific-list) for more info.
 */
export const listsUrl = (list_id?: number) =>
    list_id ? `/lists/${encodeURIComponent(list_id)}` : '/lists'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-a-specific-list-entry) for more info.
 */
export const listEntriesUrl = (list_id: number, list_entry_id?: number) =>
    listsUrl(list_id) + (
        list_entry_id
            ? `/list-entries/${encodeURIComponent(list_entry_id)}`
            : '/list-entries'
    )

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#rate-limit) for more info.
 */
export const rateLimitUrl = () => '/rate-limit'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#whoami) for more info.
 */
export const whoAmIUrl = () => '/auth/whoami'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#fields) for more info.
 */
export const fieldsUrl = (field_id?: number) =>
    field_id ? `/fields/${encodeURIComponent(field_id)}` : '/fields'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#field-values) for more info.
 */
export const fieldValuesUrl = (field_value_id?: number) => {
    return field_value_id
        ? `/field-values/${encodeURIComponent(field_value_id)}`
        : '/field-values'
}

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-field-values-changes) for more info.
 */
export const fieldValueChangesUrl = () => '/field-value-changes'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#organizations) for more info.
 */
export const organizationsUrl = (organization_id?: number | 'fields') => {
    return organization_id
        ? `/organizations/${encodeURIComponent(organization_id)}`
        : '/organizations'
}

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-global-organizations-fields) for more info.
 */
export const organizationFieldsUrl = () => organizationsUrl('fields')

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#persons) for more info.
 */
export const personsUrl = (person_id?: number | 'fields') => {
    return person_id ? `/persons/${encodeURIComponent(person_id)}` : '/persons'
}

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-global-person-fields) for more info.
 */
export const personFieldsUrl = () => personsUrl('fields')

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#entity-files) for more info.
 */
export const entityFilesUrl = (
    entity_file_id?: number,
    is_download: boolean = false,
) => {
    return entity_file_id
        ? `/entity-files` + (is_download ? '/download' : '') +
            `/${encodeURIComponent(entity_file_id)}`
        : '/entity-files'
}
