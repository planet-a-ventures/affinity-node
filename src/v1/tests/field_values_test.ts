import { assertSnapshot } from '@std/testing/snapshot'
import { afterEach, beforeEach, describe, it } from '@std/testing/bdd'
import { assertEquals } from '@std/assert'

import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import { Affinity, FieldValues } from '../index.ts'
import { getRawFixture } from './get_raw_fixture.ts'
import { apiKey, isLiveRun } from './env.ts'
import { fieldValuesUrl } from '../urls.ts'

describe('field_values', () => {
    let mock: MockAdapter
    let affinity: Affinity

    beforeEach(() => {
        if (!isLiveRun()) {
            mock = new MockAdapter(axios, { onNoMatch: 'throwException' })
        }
        affinity = new Affinity(apiKey() || 'api_key')
    })
    afterEach(() => {
        mock?.reset()
    })

    it('can get all', async (t) => {
        const params = { person_id: 38706 }
        mock?.onGet(fieldValuesUrl(), { params }).reply(
            200,
            await getRawFixture('field_values/get_all.raw.response.json'),
        )
        const res = await affinity.fieldValues.all(params)
        await assertSnapshot(t, res)
    })

    it('can create a new field value', async (t) => {
        const data = {
            field_id: 1284,
            entity_id: 38706,
            value: 'Architecture',
        }
        mock?.onPost(fieldValuesUrl()).reply(
            201,
            await getRawFixture('field_values/create.raw.response.json'),
        )
        const res = await affinity.fieldValues.create(data)
        await assertSnapshot(t, res)
    })

    it('can update a field value', async (t) => {
        const data = {
            field_value_id: 20406836,
            value: 'Healthcare',
        }
        mock?.onPut(fieldValuesUrl(data.field_value_id), {
            value: data.value,
        }).reply(
            200,
            await getRawFixture('field_values/update.raw.response.json'),
        )
        const res = await affinity.fieldValues.update(data)
        await assertSnapshot(t, res)
    })

    it('can update a field value with a date ', async (t) => {
        const data = {
            field_value_id: 123456,
            value: new Date(500000000000),
        }
        mock?.onPut(fieldValuesUrl(data.field_value_id), {
            value: data.value.toISOString(),
        }).reply(
            200,
            await getRawFixture('field_values/update2.raw.response.json'),
        )
        const res = await affinity.fieldValues.update(data)
        await assertSnapshot(t, res)
    })

    it('can delete a field value', async (t) => {
        const field_value_id = 20406836
        mock?.onDelete(fieldValuesUrl(field_value_id)).reply(200, {
            success: true,
        })
        const res = await affinity.fieldValues.delete({ field_value_id })
        assertEquals(res, true)
    })
})
