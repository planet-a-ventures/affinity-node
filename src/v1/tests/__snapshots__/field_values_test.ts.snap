export const snapshot = {};

snapshot[`field_values > can get all 1`] = `
[
  {
    created_at: 2021-10-04T12:54:24.694Z,
    entity_id: 38706,
    field_id: 337,
    id: 250616,
    list_entry_id: null,
    updated_at: 2022-03-04T12:54:24.694Z,
    value: {
      city: "San Francisco",
      continent: null,
      country: "United States",
      state: "California",
      street_address: null,
    },
  },
  {
    created_at: 2022-10-04T12:54:24.694Z,
    entity_id: 65680071,
    entity_type: 0,
    field_id: 768101,
    id: 2634897436,
    list_entry_id: null,
    updated_at: null,
    value: "Software Engineer",
    value_type: 2,
  },
]
`;

snapshot[`field_values > can create a new field value 1`] = `
{
  created_at: 2022-10-04T14:37:19.418Z,
  entity_id: 38706,
  field_id: 1284,
  id: 20406836,
  list_entry_id: null,
  updated_at: null,
  value: "Architecture",
}
`;

snapshot[`field_values > can update a field value 1`] = `
{
  created_at: 2022-10-04T14:37:19.418Z,
  entity_id: 38706,
  field_id: 1284,
  id: 20406836,
  list_entry_id: null,
  updated_at: 2023-02-06T16:53:08.914Z,
  value: "Healthcare",
}
`;
